bftengine/tests/controllerWithSimpleHistory/ControllerWithSimpleHistory_test.cpp

```diff
@@ -145,7 +147,319 @@ TEST(ControllerWithSimpleHistory, onNewView_not_primary) {
+  }
+
+  ASSERT_EQ(changed, true);
+  ASSERT_EQ(CommitPath::FAST_WITH_THRESHOLD, cwsh.getCurrentFirstPath());
+
+  // Start from next window
+  for (auto i = (size_t)65; i < 65 + ControllerWithSimpleHistory::EvaluationPeriod; ++i) {
+    cwsh.onSendingPrePrepare((SeqNum)i, CommitPath::OPTIMISTIC_FAST, prePrepareTime);
+    changed = cwsh.onNewSeqNumberExecution((SeqNum)i);
+    cwsh.onStartingSlowCommit((SeqNum)i);
+  }
+
+  ASSERT_EQ(changed, true);
+  ASSERT_EQ(CommitPath::SLOW, cwsh.getCurrentFirstPath());
+}
+
+// Test - no degradation
+// Logic:
+// 1)Construct controller with C > 0 i.e. enable FAST_WITH_THRESHOLD path.
+// 2)Run loop for EvaluationPeriod times:
+//  - call to onStartingSlowCommit for less than (EvaluationPeriod - factor * EvaluationPeriod)
+TEST(ControllerWithSimpleHistory, no_degradation) {
+  uint16_t C = 1;
+  uint16_t F = 1;
+  ReplicaId replicaId = 0;
+  ViewNum initialView = 0;
+  SeqNum initialSeq = 0;
+  ControllerWithSimpleHistory cwsh{C, F, replicaId, initialView, initialSeq};
+
+  ASSERT_EQ(CommitPath::OPTIMISTIC_FAST, cwsh.getCurrentFirstPath());
+  bool changed{true};
+
+  // Less than needed to trigger degradation
+  auto slowStartCount =
+      ControllerWithSimpleHistory::EvaluationPeriod * (1 - ControllerWithSimpleHistory_debugDowngradeFactor) - 1;
+
+  auto prePrepareTime = std::chrono::steady_clock::now() - std::chrono::milliseconds(10);
+  for (auto i = (size_t)1; i <= ControllerWithSimpleHistory::EvaluationPeriod; ++i) {
+    cwsh.onSendingPrePrepare((SeqNum)i, CommitPath::OPTIMISTIC_FAST, prePrepareTime);
+    changed = cwsh.onNewSeqNumberExecution((SeqNum)i);
+    if (slowStartCount-- > 0) cwsh.onStartingSlowCommit((SeqNum)i);
+  }
+  ASSERT_EQ(changed, false);
+  ASSERT_EQ(CommitPath::OPTIMISTIC_FAST, cwsh.getCurrentFirstPath());
+}
+
+// Upgrade scenarios
+
+// Mocks
+class IShareSecretKeyDummy : public IShareSecretKey {
+ public:
+  string toString() const override { return "IShareSecretKeyDummy"; }
+};
+
+class IShareVerificationKeyDummy : public IShareVerificationKey {
+ public:
+  string toString() const override { return "IShareVerificationKeyDummy"; }
+};
+
+class ThreshSigMock : public IThresholdSigner {
+ public:
+  IShareSecretKeyDummy is;
+  IShareVerificationKeyDummy isv;
+  virtual int requiredLengthForSignedData() const { return 5; };
+  virtual void signData(const char *hash, int hashLen, char *outSig, int outSigLen){};
+
+  virtual const IShareSecretKey &getShareSecretKey() { return is; };
+  virtual const IShareVerificationKey &getShareVerificationKey() const { return isv; };
+  const std::string getVersion() const { return "v"; }
+  void serializeDataMembers(std::ostream &) const {}
+  void deserializeDataMembers(std::istream &) {}
+  const IShareSecretKey &getShareSecretKey() const { return is; }
+};
+
+// Test - Downgrade from OPTIMISTIC_FAST to FAST_WITH_THRESHOLD then upgrade to OPTIMISTIC_FAST
+// E.L Logic: Not Possible, Bug ?
```

Comment:
is this a TODO ?

Reply:
No, I planned to do that test ,then realized it's not possible, logic does not support that behaviour, I believe it's a bug, though currently not major.